misc/proper_plugin.py

```diff
@@ -40,21 +42,40 @@ def isinstance_proper_hook(ctx: FunctionContext) -> Type:
     for arg in ctx.arg_types[0]:
         if is_improper_type(arg):
             right = get_proper_type(ctx.arg_types[1][0])
-            if isinstance(right, CallableType) and right.is_type_obj():
-                if right.type_object().fullname() in ('mypy.types.Type',
-                                                      'mypy.types.ProperType',
-                                                      'mypy.types.TypeAliasType'):
-                    # Special case: things like assert isinstance(typ, ProperType) are always OK.
-                    return ctx.default_return_type
-                if right.type_object().fullname() in ('mypy.types.UnboundType',
-                                                      'mypy.types.TypeVarType'):
-                    # Special case: these are not valid targets for a type alias and thus safe.
-                    return ctx.default_return_type
+            if is_special_target(right):
+                return ctx.default_return_type
             ctx.api.fail('Never apply isinstance() to unexpanded types;'
                          ' use mypy.types.get_proper_type() first', ctx.context)
     return ctx.default_return_type
 
 
+def is_special_target(right: ProperType) -> bool:
+    """Whitelist some special cases for use in isinstance() with improper types."""
+    if isinstance(right, CallableType) and right.is_type_obj():
+        if right.type_object().fullname() == 'builtins.tuple':
+            # Used with Union[Type, Tuple[Type, ...]].
+            return True
+        if right.type_object().fullname() in ('mypy.types.Type',
+                                              'mypy.types.ProperType',
+                                              'mypy.types.TypeAliasType'):
+            # Special case: things like assert isinstance(typ, ProperType) are always OK.
+            return True
+        if right.type_object().fullname() in ('mypy.types.UnboundType',
+                                              'mypy.types.TypeVarType',
+                                              'mypy.types.EllipsisType',
+                                              'mypy.types.StarType',
+                                              'mypy.types.TypeList',
+                                              'mypy.types.CallableArgument',
+                                              'mypy.types.PartialType',
+                                              'mypy.types.ErasedType'):
+            # Special case: these are not valid targets for a type alias and thus safe.
+            # TODO: introduce a SyntheticType base to simplify this?
+            return True
+    elif isinstance(right, TupleType):
+        return all(is_special_target(t) for t in get_proper_types(right.items))
+    return False
+
+
 def is_improper_type(typ: Type) -> bool:
```

Comment:
Should we also consider 'Any' to be an improper type? (Could be overkill/out-of-scope for this PR)

Reply:
I think this and below actually makes sense. I will try and if it is not hard, I will add to this PR.